### PR TITLE
feat: VTL util.log types

### PR DESCRIPTION
**Pull request by volkanunsal**
Marks progress towards #61

### CODE REVIEW

File: src/appsync.ts [hunk 1]
```diff
@@ -370,12 +370,12 @@ export class AppsyncResolver<
                 // lambdaFunction().prop
                 // table.query().Items.size() //.size() is still a Call but not a call to a service.
                 return expr;
-              } else if (expr.kind === "PropAccessExpr") {
+              } else if (expr.kind === 'PropAccessExpr') {
```
Comment (thantos) [inline src/appsync.ts:373]: Please ensure you are using the prettier profile with the project.

File: src/appsync.ts [hunk 2]
```diff
@@ -708,6 +708,25 @@ export interface $util {
...
+  /**
+   * Logs the string representation of the provided object to the requested log stream when field-level CloudWatch logging is enabled with log level ERROR or log level ALL on an API.
+   */
+  error(obj: Record<string, any>): void;
+  /**
+   * Logs the string representation of the provided objects to the requested log stream when field-level CloudWatch logging is enabled with log level ERROR or log level ALL on an API. This utility will replace all variables indicated by "{}" in the first input format string with the string representation of the provided objects in order.
+   */
+  error(message: string, rest: Record<string, any>[]): void;
+}
+
```
Comment (thantos) [inline src/appsync.ts:736]: Awesome! the recent utils additions to app sync are exciting to see.

File: src/appsync.ts [hunk 3]
```diff
@@ -708,6 +715,25 @@ export interface $util {
...
+   */
+  info(message: string, rest: Record<string, any>[]): void;
+  /**
+   * Logs the string representation of the provided object to the requested log stream when field-level CloudWatch logging is enabled with log level ERROR or log level ALL on an API.
+   */
+  error(obj: Record<string, any>): void;
+  /**
+   * Logs the string representation of the provided objects to the requested log stream when field-level CloudWatch logging is enabled with log level ERROR or log level ALL on an API. This utility will replace all variables indicated by "{}" in the first input format string with the string representation of the provided objects in order.
+   */
+  error(message: string, rest: Record<string, any>[]): void;
```
Comment (thantos) [inline src/appsync.ts:734]: `...rest: Record<string, any>[]`

**Comment by volkanunsal** [comment]
Sorry about all the formatting changes. My prettier seems to be using a different config than yours. Maybe it'd be a good idea to include a prettier config for the project.

**Review by thantos — CHANGES_REQUESTED** [review]
Please add tests. I just merged a change to use snapshot tests and an updated appSyncTest method for app sync/vtl that'll make testing app sync easier.

**Comment by thantos** [comment]
> Sorry about all the formatting changes. My prettier seems to be using a different config than yours. Maybe it'd be a good idea to include a prettier config for the project.

Hmm.

Could have sworn we had one, will look into it. (not sure how it was working up until now).

**Comment by volkanunsal** [comment]
There is a git commit hook that runs prettier, which uses an internal config probably.

**Comment by volkanunsal** [comment]
Should be good now.

File: test/vtl.test.ts [hunk 4]
```diff
@@ -359,6 +359,34 @@ test("$util.time.nowISO8601", () =>
     })
   ));
 
+test("$util.log.info(message)", () =>
+  appsyncTestCase(
+    reflect(() => {
+      return $util.log.info("hello world");
+    })
+  ));
```
Comment (sam-goodwin) [inline test/vtl.test.ts:367]: Does this create snapshot tests? I don't see them in the PR

File: src/appsync.ts [hunk 5]
```diff
@@ -708,6 +715,25 @@ export interface $util {
...
+   */
+  info(obj: Record<string, any>): void;
+  /**
+   * Logs the string representation of the provided objects to the requested log stream when request-level and field-level CloudWatch logging is enabled with log level ALL on an API. This utility will replace all variables indicated by "{}" in the first input format string with the string representation of the provided objects in order.
+   */
+  info(message: string, ...rest: Record<string, any>[]): void;
+  /**
+   * Logs the string representation of the provided object to the requested log stream when field-level CloudWatch logging is enabled with log level ERROR or log level ALL on an API.
+   */
+  error(obj: Record<string, any>): void;
```
Comment (sam-goodwin) [inline src/appsync.ts:730]: Is it required to be an object? Can it be a primitive like string?

**Review by sam-goodwin — CHANGES_REQUESTED** [review]

File: test/vtl.test.ts (hunk 4 above)
Comment (volkanunsal) [reply test/vtl.test.ts:367]: I can't run the tests locally, for some reason. It fails due to AWS account number not being valid. Is there another command to record the snapshots?

File: src/appsync.ts (hunk 5 above)
Comment (volkanunsal) [reply src/appsync.ts:730]: you're right. it's a string.

Comment (volkanunsal) [reply src/appsync.ts:730]: ah, actually, I was wrong... it does say it's an object.

> $util.log.info(Object) : Void

Comment (volkanunsal) [reply src/appsync.ts:730]: The second method signature handles the string usecase, though.

File: test/vtl.test.ts (hunk 4 above)
Comment (sam-goodwin) [reply test/vtl.test.ts:367]: Oh have you set up local stack? Our tests spin that up locally and deploy stacks to it. https://localstack.cloud/

We need to set up a contributing guide. Thanks for the reminder.

Comment (sam-goodwin) [reply test/vtl.test.ts:367]: Tracking https://github.com/functionless/functionless/issues/234

File: src/appsync.ts (hunk 5 above)
Comment (sam-goodwin) [reply src/appsync.ts:730]: AWS's documentation uses `Object` as `any` - it's Java based, so `Object``` just means any value type can be passed in. Not `Record<string, any>`

**Comment by sam-goodwin** [comment]
Thanks for this change @volkanunsal! Super useful feature. I wonder if we should compile console.log to $util.log?

Also, a suggestion for your development lifecycle. Don't use `git push --force`, that incremental changes are visible as individual commits on the PR. When we merge all the commits are squashed into one, so everything is clean.

**Comment by thantos** [comment]
Something else we should keep in mind is that logging in appSync requires logging to be enabled. 

Should we:
1. validate that logging is enabled and fail if logging is used, but not enabled?
2. Turn on logging when logging is used? probably not just a flag.
3. Nothing, appSync doesn't describe the ramification of using `log` when logging is off, so I assume it just does nothing?

File: src/appsync.ts (hunk 5 above)
Comment (thantos) [reply src/appsync.ts]: Confirmed

```
$util.log.info(" 1 {} 2 {}", $v1.containsKey('key'), "hi")
```

out

```
c98f529-d1b3-4a03-a5ff-22853b3caa4e INFO - 1 true 2 hi
```

Comment (thantos) [reply src/appsync.ts]: Note, anything in the first position when more than 1 argument is given appears to fail the log... (I tried passing a boolean).

**Comment by sam-goodwin** [comment]
> Nothing, appSync doesn't describe the ramification of using log when logging is off, so I assume it just does nothing?

Test this. If true, then I think this should be the behavior.

**Comment by volkanunsal** [comment]
FYI. I couldn't run the tests with `project test`, and had to use `project build` instead.

File: test/__snapshots__/vtl.test.ts.snap [hunk 6]
```diff
@@ -1,5 +1,53 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`$util.log.error(message) 1`] = `
+Array [
+  "{
+  \\"version\\": \\"2018-05-29\\",
+  \\"payload\\": null
+}",
+  "#return($util.log.error('hello world'))",
```
Comment (sam-goodwin) [inline test/__snapshots__/vtl.test.ts.snap:9]: Curious what this does at runtime? Return null?

**Review by sam-goodwin — APPROVED** [review]

**Pull request merged**